File: src/utils/git.ts
/**
 * Git Utilities — Enhanced with GitFlow Standards
 *
 * Provides GitFlow branching, merge policy enforcement,
 * and safe rollback capabilities for Freddie Mac standards.
 */

import simpleGit from "simple-git";
import { paths } from "../config/paths";
import { Ticket } from "../core/types";
import {
  buildBranchName,
  buildCommitMessage,
  validateBranchName,
  validateCommitMessage,
} from "./git-policy";

const git = simpleGit(paths.rootDir);

// ---------------------------------------------------------------------------
// Original functionality (preserved)
// ---------------------------------------------------------------------------

export async function pushTicket(ticket: Ticket): Promise<string> {
  try {
    const branchName = buildBranchName(ticket.id, "feature");
    const commitMsg = buildCommitMessage(ticket.id, ticket.title, "feat");

    // Validate branch name against policy
    const branchViolations = validateBranchName(branchName);
    if (branchViolations.length > 0) {
      return `Policy violation: ${branchViolations.map((v) => v.message).join("; ")}`;
    }

    // Validate commit message against policy
    const commitViolations = validateCommitMessage(commitMsg);
    if (commitViolations.some((v) => v.severity === "error")) {
      return `Commit policy violation: ${commitViolations.map((v) => v.message).join("; ")}`;
    }

    const branches = await git.branchLocal();

    if (!branches.all.includes(branchName)) {
      await git.checkoutLocalBranch(branchName);
    } else {
      await git.checkout(branchName);
    }

    await git.add(".");
    await git.commit(commitMsg);
    await git.push("origin", branchName, ["-u"]);
    return `Pushed changes to origin/${branchName}.`;
  } catch (error) {
    return `Skipped push: ${formatGitError(error)}`;
  }
}

export async function getChangedFiles(): Promise<string[]> {
  try {
    const status = await git.status();
    return [
      ...status.created,
      ...status.modified,
      ...status.deleted,
      ...status.not_added,
      ...status.renamed.map((entry) => `${entry.from} -> ${entry.to}`),
    ];
  } catch (error) {
    return [`Git unavailable: ${formatGitError(error)}`];
  }
}

// ---------------------------------------------------------------------------
// GitFlow — Branching
// ---------------------------------------------------------------------------

/**
 * Create a feature branch from develop (or current branch).
 */
export async function createFeatureBranch(ticketId: string): Promise<string> {
  try {
    const branchName = buildBranchName(ticketId, "feature");
    const violations = validateBranchName(branchName);
    if (violations.length > 0) {
      return `Policy violation: ${violations.map((v) => v.message).join("; ")}`;
    }

    const branches = await git.branchLocal();
    if (branches.all.includes(branchName)) {
      await git.checkout(branchName);
      return `Switched to existing branch '${branchName}'.`;
    }

    // Try to branch from develop if it exists
    if (branches.all.includes("develop")) {
      await git.checkout("develop");
    }

    await git.checkoutLocalBranch(branchName);
    return `Created and switched to branch '${branchName}'.`;
  } catch (error) {
    return `Failed to create branch: ${formatGitError(error)}`;
  }
}

/**
 * Create a release branch from develop.
 */
export async function createReleaseBranch(version: string): Promise<string> {
  try {
    const branchName = `release/${version}`;
    const violations = validateBranchName(branchName);
    if (violations.length > 0) {
      return `Policy violation: ${violations.map((v) => v.message).join("; ")}`;
    }

    const branches = await git.branchLocal();
    if (branches.all.includes("develop")) {
      await git.checkout("develop");
    }

    await git.checkoutLocalBranch(branchName);
    return `Created release branch '${branchName}' from develop.`;
  } catch (error) {
    return `Failed to create release branch: ${formatGitError(error)}`;
  }
}

/**
 * Create a hotfix branch from main.
 */
export async function createHotfixBranch(ticketId: string): Promise<string> {
  try {
    const branchName = buildBranchName(ticketId, "hotfix");
    const branches = await git.branchLocal();

    if (branches.all.includes("main")) {
      await git.checkout("main");
    }

    await git.checkoutLocalBranch(branchName);
    return `Created hotfix branch '${branchName}' from main.`;
  } catch (error) {
    return `Failed to create hotfix branch: ${formatGitError(error)}`;
  }
}

// ---------------------------------------------------------------------------
// GitFlow — Merging (with --no-ff to preserve history)
// ---------------------------------------------------------------------------

/**
 * Merge a feature branch into develop using --no-ff.
 */
export async function mergeFeatureToDevelop(ticketId: string): Promise<string> {
  try {
    const branchName = buildBranchName(ticketId, "feature");
    const branches = await git.branchLocal();

    if (!branches.all.includes(branchName)) {
      return `Branch '${branchName}' does not exist.`;
    }

    // Ensure develop exists
    const targetBranch = branches.all.includes("develop")
      ? "develop"
      : branches.current;
    await git.checkout(targetBranch);
    await git.merge([
      branchName,
      "--no-ff",
      "-m",
      `merge: ${branchName} into ${targetBranch}`,
    ]);

    return `Merged '${branchName}' into '${targetBranch}' with --no-ff (merge commit preserved).`;
  } catch (error) {
    return `Merge failed: ${formatGitError(error)}`;
  }
}

/**
 * Merge a release branch into both main and develop.
 */
export async function mergeReleaseBranch(version: string): Promise<string> {
  try {
    const branchName = `release/${version}`;
    const branches = await git.branchLocal();
    const results: string[] = [];

    if (!branches.all.includes(branchName)) {
      return `Release branch '${branchName}' does not exist.`;
    }

    // Merge into main
    if (branches.all.includes("main")) {
      await git.checkout("main");
      await git.merge([
        branchName,
        "--no-ff",
        "-m",
        `release: merge ${version} into main`,
      ]);
      results.push(`Merged '${branchName}' into main.`);
    }

    // Merge into develop
    if (branches.all.includes("develop")) {
      await git.checkout("develop");
      await git.merge([
        branchName,
        "--no-ff",
        "-m",
        `release: merge ${version} into develop`,
      ]);
      results.push(`Merged '${branchName}' into develop.`);
    }

    return results.join(" ");
  } catch (error) {
    return `Release merge failed: ${formatGitError(error)}`;
  }
}

// ---------------------------------------------------------------------------
// Rollback (non-destructive, preserves audit trail)
// ---------------------------------------------------------------------------

/**
 * Revert the last commit (safe, creates a new reversal commit).
 */
export async function rollbackLastCommit(): Promise<string> {
  try {
    await git.revert("HEAD", ["--no-edit"]);
    return "Reverted the last commit. A new reversal commit was created (audit trail preserved).";
  } catch (error) {
    return `Rollback failed: ${formatGitError(error)}`;
  }
}

/**
 * Revert to a specific commit by SHA.
 */
export async function rollbackToCommit(sha: string): Promise<string> {
  try {
    await git.revert(sha, ["--no-edit"]);
    return `Reverted commit ${sha.slice(0, 8)}. A new reversal commit was created.`;
  } catch (error) {
    return `Rollback failed: ${formatGitError(error)}`;
  }
}

/**
 * List recent commits for rollback selection.
 */
export async function listRecentCommits(count: number = 10): Promise<string[]> {
  try {
    const log = await git.log({ maxCount: count });
    return log.all.map((entry) => {
      const short = entry.hash.slice(0, 8);
      const date = entry.date.split("T")[0];
      return `${short} | ${date} | ${entry.message}`;
    });
  } catch (error) {
    return [`Unable to read git log: ${formatGitError(error)}`];
  }
}

/**
 * Get current branch name.
 */
export async function getCurrentBranch(): Promise<string> {
  try {
    const branches = await git.branchLocal();
    return branches.current;
  } catch {
    return "unknown";
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function formatGitError(error: unknown): string {
  if (error instanceof Error) {
    if (error.message.includes("spawn git ENOENT")) {
      return "git is not installed or not available on PATH";
    }

    return error.message;
  }

  return "git is not available";
}

File: src/utils/git.py
"""Git utilities — replaces git.ts. Uses GitPython for all Git operations."""
from __future__ import annotations
from git import Repo, InvalidGitRepositoryError
from ..config.paths import paths
from ..core.types import Ticket
from .git_policy import build_branch_name, build_commit_message, validate_branch_name, validate_commit_message

def _get_repo() -> Repo:
    return Repo(str(paths["root_dir"]))

def _fmt_err(e: Exception) -> str:
    msg = str(e)
    if "git" in msg.lower() and "not found" in msg.lower(): return "git is not installed or not on PATH"
    return msg

# ── Push Ticket ──────────────────────────────────────────────
async def push_ticket(ticket: Ticket) -> str:
    try:
        repo = _get_repo()
        branch = build_branch_name(ticket.id, "feature")
        commit_msg = build_commit_message(ticket.id, ticket.title, "feat")
        bv = validate_branch_name(branch)
        if bv: return f"Policy violation: {'; '.join(v.message for v in bv)}"
        cv = validate_commit_message(commit_msg)
        if any(v.severity == "error" for v in cv): return f"Commit policy violation: {'; '.join(v.message for v in cv)}"
        if branch not in [h.name for h in repo.heads]:
            repo.create_head(branch).checkout()
        else:
            repo.heads[branch].checkout()
        repo.git.add(".")
        repo.index.commit(commit_msg)
        repo.git.push("origin", branch, "-u")
        return f"Pushed changes to origin/{branch}."
    except Exception as e: return f"Skipped push: {_fmt_err(e)}"

async def get_changed_files() -> list[str]:
    try:
        repo = _get_repo()
        s = repo.git.status("--porcelain")
        return [line.strip().split(None, 1)[-1] for line in s.splitlines() if line.strip()] if s.strip() else []
    except Exception as e: return [f"Git unavailable: {_fmt_err(e)}"]

async def create_feature_branch(ticket_id: str) -> str:
    try:
        repo = _get_repo()
        branch = build_branch_name(ticket_id, "feature")
        if branch in [h.name for h in repo.heads]:
            repo.heads[branch].checkout(); return f"Switched to existing branch '{branch}'."
        if "develop" in [h.name for h in repo.heads]: repo.heads.develop.checkout()
        repo.create_head(branch).checkout()
        return f"Created and switched to branch '{branch}'."
    except Exception as e: return f"Failed: {_fmt_err(e)}"

async def create_release_branch(version: str) -> str:
    try:
        repo = _get_repo()
        branch = f"release/{version}"
        if "develop" in [h.name for h in repo.heads]: repo.heads.develop.checkout()
        repo.create_head(branch).checkout()
        return f"Created release branch '{branch}'."
    except Exception as e: return f"Failed: {_fmt_err(e)}"

async def create_hotfix_branch(ticket_id: str) -> str:
    try:
        repo = _get_repo()
        branch = build_branch_name(ticket_id, "hotfix")
        if "main" in [h.name for h in repo.heads]: repo.heads.main.checkout()
        repo.create_head(branch).checkout()
        return f"Created hotfix branch '{branch}'."
    except Exception as e: return f"Failed: {_fmt_err(e)}"

async def merge_feature_to_develop(ticket_id: str) -> str:
    try:
        repo = _get_repo()
        branch = build_branch_name(ticket_id, "feature")
        if branch not in [h.name for h in repo.heads]: return f"Branch '{branch}' does not exist."
        target = "develop" if "develop" in [h.name for h in repo.heads] else repo.active_branch.name
        repo.heads[target].checkout()
        repo.git.merge(branch, "--no-ff", "-m", f"merge: {branch} into {target}")
        return f"Merged '{branch}' into '{target}' with --no-ff."
    except Exception as e: return f"Merge failed: {_fmt_err(e)}"

async def rollback_last_commit() -> str:
    try:
        _get_repo().git.revert("HEAD", "--no-edit")
        return "Reverted last commit. Reversal commit created (audit trail preserved)."
    except Exception as e: return f"Rollback failed: {_fmt_err(e)}"

async def rollback_to_commit(sha: str) -> str:
    try:
        _get_repo().git.revert(sha, "--no-edit")
        return f"Reverted commit {sha[:8]}."
    except Exception as e: return f"Rollback failed: {_fmt_err(e)}"

async def list_recent_commits(count: int = 10) -> list[str]:
    try:
        repo = _get_repo()
        return [f"{c.hexsha[:8]} | {c.committed_datetime.strftime('%Y-%m-%d')} | {c.summary}" for c in repo.iter_commits(max_count=count)]
    except Exception as e: return [f"Unable to read git log: {_fmt_err(e)}"]

async def get_current_branch() -> str:
    try: return str(_get_repo().active_branch)
    except Exception: return "unknown"
